corehq/apps/domain/tests/test_auth.py

```diff
@@ -64,3 +66,183 @@ def test_request_for_valid_domain_membership_for_non_superuser(self, *args):
+        self._create_api_key_for_user(user, key='1234', is_active=False)
+
+        request = self._create_request()
+
+        self.assertIsNone(self.backend.authenticate(request, 'test@dimagi.com', '1234'))
+
+    def test_allows_access_via_domain_restricted_keys(self):
+        user = self._create_user('test@dimagi.com')
+        self._create_api_key_for_user(user, key='1234', domain='test-domain')
+
+        request = self._create_request(for_domain='test-domain')
+
+        self.assertEqual(self.backend.authenticate(request, 'test@dimagi.com', '1234'), user)
+
+    def test_does_not_allow_cross_domain_access(self):
+        user = self._create_user('test@dimagi.com')
+        self._create_api_key_for_user(user, key='1234', domain='test-domain1')
+
+        request = self._create_request(for_domain='test-domain2')
+
+        self.assertIsNone(self.backend.authenticate(request, 'test@dimagi.com', '1234'))
+
+    def test_allows_access_to_ips_on_whitelist(self):
+        user = self._create_user('test@dimagi.com')
+        self._create_api_key_for_user(user, key='1234', allowed_ips=['127.0.0.1'])
+
+        request = self._create_request(for_domain='test-domain2', ip='127.0.0.1')
+
+        self.assertEqual(self.backend.authenticate(request, 'test@dimagi.com', '1234'), user)
+
+    def test_does_not_allow_ips_not_on_whitelist(self):
+        user = self._create_user('test@dimagi.com')
+        self._create_api_key_for_user(user, key='1234', allowed_ips=['127.0.0.1'])
+
+        request = self._create_request(for_domain='test-domain2', ip='5.5.5.5')
+
+        self.assertIsNone(self.backend.authenticate(request, 'test@dimagi.com', '1234'))
+
+    def setUp(self):
+        self.backend = ApiKeyFallbackBackend()
+
+    @classmethod
+    def _create_user(cls, username):
+        return User.objects.create_user(username=username, password='password')
+
+    @classmethod
+    def _create_api_key_for_user(
+            cls, user, name='ApiKey', key='1234', is_active=True, allowed_ips=None, domain=''):
+        allowed_ips = allowed_ips or []
+        return HQApiKey.objects.create(
+            user=user, key=key, name=name, ip_allowlist=allowed_ips, domain=domain, is_active=is_active)
+
+    @classmethod
+    def _create_request(cls, can_use_api_key=True, for_domain='test-domain', ip=None):
+        request = HttpRequest()
+        request.domain = for_domain
+
+        if ip:
+            request.META['HTTP_X_FORWARDED_FOR'] = ip
+
+        if can_use_api_key:
+            request.check_for_password_as_api_key = True
+
+        return request
+
+
+class HQApiKeyAuthenticationTests(TestCase):
+    def test_successful_attempt_returns_true(self):
+        user = self._create_user('test@dimagi.com')
+        self._create_api_key_for_user(user, key='1234')
+        request = self._create_request(username='test@dimagi.com', api_key='1234')
+
+        auth = HQApiKeyAuthentication()
+
+        self.assertTrue(auth.is_authenticated(request))
```

Comment:
`asssertTrue(obj)` does something like `assert bool(obj)`. Could this test pass when the returned object is a response with `status_code == 401`?

Reply:
Thanks, definitely missed this. Those assertions were broken and were passing even on errors. Changed those assertions to use `assertIs(..., True)` instead.